### DISCUSSION TITLE
Feature requests

### DISCUSSION THREAD

**Discussion by hrvach** [hrvach/deskhop#27] [Ideas]
This thread is for feature requests.

I've tried compiling the issues mentioned before. If your suggestion is not on the list, feel free to add it to the comment.

- More than 2 devices
- Support for PS/2 devices
- Windows 10+ multi-display workaround
- Support for variable/configurable speed per-display
- Better support with only 1 device connected
- Bluetooth support

PCB updates:

- Dot marking pin 1
- Holes on pads to support other boards
- Silkscreen mark which pins have to be soldered

**Comment by seii-m** [comment]
I would like you to add a macro function.
Or please tell me which part of the source code I need to modify to achieve this.

**Reply by hrvach** [reply]
Hi, what kind of macros did you have in mind, keyboard? Any examples?

**Reply by seii-m** [reply]
I just want to assign different functions to the keys on my unused keyboard.

**Comment by hrvach** [comment]
You can check the beginning of keyboard.c and add a custom handler to remap keys. If you need to map multiple keys, that's a bit more difficult. I will look into this, but some other features are more pressing so no timeline estimates.

**Reply by seii-m** [reply]
Thanks! I got it!!
By the way, Is there a way to make a multi-button mouse compatible?

**Comment by p3lim** [comment]
Would also be interested in macro support, with these options:
- which keys to activate (modifiers and keys)
- which keys to send to host (which could be separate from the keys to activate)
- which host(s) to send to

With this I could make a macro to send super+L to lock both hosts.

**Reply by p3lim** [reply]
This might be a bit much for user_config.h though :smile:.

**Reply by p3lim** [reply]
My initial attempt at this was this (which didn't work):

```diff
diff --git a/src/keyboard.c b/src/keyboard.c
index 3a59f8a..90fe813 100644
--- a/src/keyboard.c
+++ b/src/keyboard.c
@@ -77,7 +77,17 @@ hotkey_combo_t hotkeys[] = {
      .keys           = {HID_KEY_B},
      .key_count      = 1,
      .acknowledge    = true,
-     .action_handler = &fw_upgrade_hotkey_handler_B}};
+     .action_handler = &fw_upgrade_hotkey_handler_B},
+
+    {.modifier       = KEYBOARD_MODIFIER_LEFTGUI,
+     .keys           = {HID_KEY_L},
+     .key_count      = 1,
+     .pass_to_os     = true},
+
+    {.modifier       = KEYBOARD_MODIFIER_RIGHTGUI,
+     .keys           = {HID_KEY_L},
+     .key_count      = 1,
+     .pass_to_os     = true}};
 
 /* ============================================================ *
  * Detect if any hotkeys were pressed
@@ -189,7 +199,13 @@ void process_keyboard_report(uint8_t *raw_report, int length, device_t *state) {
             blink_led(state);
 
         /* Execute the corresponding handler */
-        hotkey->action_handler(state);
+        if (hotkey->action_handler != NULL)
+            hotkey->action_handler(state);
+        else {
+            queue_kbd_report(keyboard_report, state);
+            send_packet((uint8_t *)keyboard_report, KEYBOARD_REPORT_MSG, KBD_REPORT_LENGTH);
+            return;
+        }
 
         /* And pass the key to the output PC if configured to do so. */
         if (!hotkey->pass_to_os)

```

**Comment by hrvach** [comment]
> With this I could make a macro to send super+L to lock both hosts.

This is a *great* idea, btw! :) 

I added the OS definition for the latest version so this enables the box to know what to send to lock a specific output.
If you have more good ideas, please share :-)

**Comment by PhetaFox** [comment]
I know there's currently limited support for multi-displays and multi-devices anyway, but I'm wondering if this can completely turn off the switching function if desired. I have 4 monitors, and sometimes I'll swap video inputs on the 3rd or 4th ones to a different system running MacOS or SteamOS. But during normal operating hours I'm just using all 4 monitors on Windows. I know there's the Lock/Unlock feature, but I'm unclear on if the possibility of a monitor to be different systems at different times breaks that functionality.

To clarify, I'm requesting that, when selected, no switching happens at all. Then, when selected, the movement to a monitor of my choice triggers a switch to the pc of my choice.

I hope that made sense. Thanks! And in agreeance with the other thread, I'd also love to buy you a coffee :)

**Comment by hrvach** [comment]
> To clarify, I'm requesting that, when selected, no switching happens at all. Then, when selected, the movement to a monitor of my choice triggers a switch to the pc of my choice.

![image](https://github.com/hrvach/deskhop/assets/7192749/a5a9bcdf-9057-4c8a-877b-e817155bf5d5)

I'm really trying to understand this and help but I fail short. Please ELI5 - how many computers, how many monitors per computer, what operating systems and what it is exactly you're trying to achieve? 😃

**Reply by PhetaFox** [reply]
No worries, I know I suck at explaining😬
I have 4 monitors, 1 Windows machine, 1 Mac mini, and 1 mouse and keyboard. (Lets just forget entirely about the SteamOS I mentioned, unimportant). 90% of the time, I'm just using my Windows pc, doing whatever, all 4 monitors, livin' life. Sometimes, I will turn on my Mac mini and swap the input of my 3rd monitor so that instead of showing windows, it's showing macOS. As far as Windows is concerned, it's still my 3rd monitor, but visually, it's now macOS's only monitor.

Here's a graphic:
![Scenarios](https://github.com/hrvach/deskhop/assets/54605410/28e66832-bcdc-4b8d-a1ff-6bd4449cd0ac)

While utilizing Scenario 1, I would like DeskHop to simply pass my inputs through to the Windows pc. (The Mac may or may not be on, but I'm not using it at the time and I'd like my 3rd monitor back.)
However,
While utilizing Scenario 2, I would like DeskHop to change keyboard/mouse inputs from Windows to Mac when my mouse enters that 3rd monitor (and back again when it leaves).

I would like both scenarios to be viable, and to have the ability to swap between them. Maybe this is already implemented and I'm just barking at the wind, but I figured that one monitor maybe sometimes being a different computer would break things a little.
Thanks for reading😂.

**Comment by thinhdnn** [comment]
[Feature requests] Add a storage (SD card ... )  and share between 2 PCs.

**Reply by hrvach** [reply]
This is not added by design, since the idea is to not allow any information flow between the outputs in any scenario. What would you use if for, like copying files to it on one and then reading on another?

**Reply by thinhdnn** [reply]
Yes, my idea is to copy a file from Mac and read it on Windows. Currently, I must copy to USB and plug the USB into Windows.

**Reply by chris-010** [reply]
Or at least text-only clipboard sharing, configurable to enable/disable it either in code or via [text-based configuration management](https://github.com/hrvach/deskhop/issues/91).

**Reply by lyjpla** [reply]
I also need the simplest clipboard-sharing functionality, which could be enabled or disabled through settings (default: disabled). This wouldn't go too much against isolation principles either.

**Comment by hrvach** [comment]
> I would like both scenarios to be viable, and to have the ability to swap between them. Maybe this is already implemented and I'm just barking at the wind, but I figured that one monitor maybe sometimes being a different computer would break things a little. Thanks for reading😂.

Sorry for getting back to you so late,  I think I understand now. This is doable but pretty hard, since windows has a limitation of absolute coordinates on main screen only, and other screens are position-guesswork. 

I'd advise moving the screen where you connect the mac to the very end and putting the main windows screen right next to it. That way, it should pretty much do what you need. I only need to add the heartbeat feature finally so it blocks hopping when an output is not properly connected and enumerated (i.e. computer off).

**Comment by hrvach** [comment]
Both valid suggestions, I'll try to look into it - it would indeed be much easier having a graphical way to position and rearrange screens.

**Comment by tomjohnstone** [comment]
Bitlocker login support: https://github.com/hrvach/deskhop/discussions/122

**Comment by azwillnj** [comment]
Feature Request: Configurable option to forward all media and volume commands to one of the connected computers. 

I am using this on my personal and work computers and the personal computer is where the sound system is connected and where I play all media. If you could select a computer in the web configurator that would always receive the media commands that would be very handy.

**Comment by hrvach** [comment]
Feature request: enable having multiple shortcuts for a single functionality because some keyboards have problems accessing e.g. configuration menu

**Comment by omelettedufromagee** [comment]
Feature request: Is it feasible to detect when Device B is shutdown and lock the mouse to Device A without the need to enter gaming mode?
And then detect when Device B is back up and activate roaming again.

**Reply by hrvach** [reply]
Yes, half of it is already implemented... taking some family time so might be a tad slow but will try to implement it as soon as possible!

**Reply by omelettedufromagee** [reply]
Enjoy the holidays!

**Reply by azwillnj** [reply]
How would this work with only one computer plugged in? Because of the way the board is designed the device connected to that side loses power. Is there a way to have both input USB's powered if only one computer is plugged in?

**Comment by fabien88fr14** [comment]
Feature request: Support for Multiple Configurations.

My current setup is as follows:
- Mac OS with two screens on the left
- Windows with one screen on the right
Sometimes, I prefer to swap the configuration. For instance, when I have a lot of work to do on my Windows workstation, I switch to:
- One screen on Mac OS
- Two screens on Windows
It would be fantastic to have a feature that allows users to switch between configurations using a keyboard shortcut.

**Reply by bengrimm** [reply]
It may also be nice to add hotkeys to change the OS of the active screen. e.g. Right Shift + F12 + (L, M, or W) to change the active screen to Linux, MacOS, or Windows respectively.

**Reply by steveh** [reply]
I'd really love this too, so I can have my work PC have the big screen during the work day, and personal at other times. 
I made a stab at it here: https://github.com/hrvach/deskhop/compare/main...steveh:deskhop:toggle?expand=1 but it didn't work. I don't really know C but if @hrvach can steer me in the right direction I'll happily keep hacking on it!

**Comment by jameshpark** [comment]
Curious if it makes sense to add a user config parameter--configurable through config.htm web interface--that enables/disables output switching via mouse pointer moving across ScreenA1 | ScreenB1 boundary. My setup/arrangement is not static, and also includes vertically stacked monitors, which makes using output switching with pointer movement impractical in its current state.

I have modified my fork to add a `DISABLE_MOUSE_OUTPUT_SWITCHING` config field (disabled by default i.e. behaves just like the current implementation) that, when enabled, makes all pointer reports `RELATIVE` deltaX-deltaY values and bypasses all screen boundary detection/bounding logic.
https://github.com/jameshpark/deskhop/commit/2488c207af7fdb96b13759a4d5f278bd4c1382b6

This makes it such that when `DISABLE_MOUSE_OUTPUT_SWITCHING` == `1`/true, the mouse input is like a traditional/dumb mouse. We don't need to worry about the pointer's position and the pointer is bound to the desktops of the currently active output. The only way to switch outputs, then, is via the keyboard shortcut.

This just happens to align with my use-case/workflow and may contradict the goals of this project, but I wanted to ask the question anyway and see if anyone else has this sort of need. 😃 

P.S. @hrvach this project has addressed a unique problem that I couldn't find a solution for and I'm so glad I stumbled upon it. Great work!

**Reply by hrvach** [reply]
Have you tried gaming mode? Sounds like exactly what you're looking for (keyboard shortcut switching + relative mode mouse).

**Reply by jameshpark** [reply]
Is there a way to boot up in gaming mode as the default? If so, then yes, that's exactly what I'm looking for!

**Comment by azwillnj** [comment]
Request: Hardware and software support for 3 computers.

You have this listed in post one but I didn't see a dedicated comment for it.

**Comment by DaveW76** [comment]
The device I would find most useful is for 2 computers, one keyboard, one mouse, but designed so that it would still work even if either one or the other computer is powered off.

**Reply by kpfleming** [reply]
That can be accomplished today using a "USB power injector" cable which ensures that there is 5V power supplied to the DeskHop even if the attached computer is powered off (or there is no computer attached).

**Reply by Gorillix** [reply]
> That can be accomplished today using a "USB power injector" cable which ensures that there is 5V power supplied to the DeskHop even if the attached computer is powered off (or there is no computer attached).

Do you have any reference product for that. I legit can only find the power blockers that I used for my 3d printers but not injectors.  I have deskhop JR variant, with USB C instead of A and couldn't find a thing.

**Reply by chri2** [reply]
I'm using my deskhop to [share mouse/keyboard between my notebook and a Librem5 mobile phone](https://chrichri.ween.de/o/5889b14d5bd145e1a4bf8e1d2d26f333). I also referenced the setup in [this issue](https://github.com/hrvach/deskhop/issues/58).

Long story short: connect the device you want to take away without cutting the power to your deskhop through a usb-hub.

I'm using a usb-c dock bought at [Purism](https://shop.puri.sm/shop/usb-c-hub/), but any hub with external power should work.

**Comment by alterwen** [comment]
Feature request: "pac-man mode" 
- crossing a computer boundary (both left and right) teleports the mouse to the other computer (on the opposite border)
- allow same behavior for vertical switch

(I'm loving the gadget)

**Comment by pyinto** [comment]
It would be great to have an option to extend the build with an extra microcontroller and to have 2+ host devices as an option!

**Comment by ct511** [comment]
Wondering if we could implement the following:
Description: Using a Mac laptop and windows desktop. Using Mx master 3s mouse and mx master mini keyboard. Logitech plus software installed in both the computers. Devices connected via Logi Bolt USB receiver. 

1) Logitech options plus application could not identify the mouse and keyboard being connected. If it identifies many customisable configurations could come into action.
2) Mx master 3s:
  a) Back and Forward buttons works in the windows but not working in the MacOS.
  b) Gesture control button is not working in both the OS. The main feature request would be to use the gesture control button to          switch between virtual desktops in Windows-OS or switch between virtual desktops and windows(applications) in MacOS.
3) Mx master Mini keyboard: (or other keyboards):
  a) A provision for assigning key inputs individually for each devices; For example, in windows normal keyboard layout, ALT-key is next to the spacebar. At the same time in Mac keyboard layout, it is the command key. So when you interchangeably use devices but with the same keyboard, always a confusion occurs. So, when the Deskhop switches to MacOS device and if we can assign ALT-key = Command button under that device, this can help with very fast shortcut usages and touch typing experiences. Automatically, when the Deskhop switches back to the Windows device, as we have NOT assigned the changes under that device, it takes in the usual ALT-key input only.

Hope, i was able to explain. Many thanks.

**Comment by mglushko** [comment]
Can we have a feature like in Synergy where you have to double tap the edge of one monitor before it switches to the other?

<img width="408" height="176" alt="image" src="https://github.com/user-attachments/assets/03a3852e-e766-414a-9f13-3090d354242c" />

**Comment by sevki** [comment]
Single monitor support with picture in picture

**Comment by bryanforbes** [comment]
A few quality of life options that I'd like to see:

- The ability to turn off the LED indicator after a (customizable) period of time of inactivity and/or after a (customizable) period after switching screens. My office is in my bedroom and the LED indicator is bright at night! But it also might be nice to just flash the LED or only have it on for a few seconds after switching.
- Shortcut customization. I'm on a Glove80 with a custom layout and it's difficult to do some of these shortcuts (F12 is the bane of my existence!). Yes, I could add them to a layer in my Glove80 config as a one-key combo but others may like to customize the shortcuts.
- A more user-friendly input unit for the screensaver idle time. I imagine most users are going to be using minutes or seconds. Having a drop-down with selectable units would be fine, but microseconds really aren't something most people think in.
- Tooltips or help text for some of the more obscure options in the configurator

**Reply by mglushko** [reply]
[#364](https://github.com/hrvach/deskhop/pull/364) turns the screensaver timers into seconds. The LED timeout and configurable shortcuts are running in a build at [deskhop-extended](https://github.com/mglushko/deskhop-extended) if you want them before they are added here. I made it as a personal build with some features I wanted.

**Comment by ak66666** [comment]
Wrap-around cursor movement option: when the cursor reaches external outer edge to jump to the opposite edge of the other computer's screen, so it goes in the same direction over and over. I think that would be handy with the trackball - I often find myself navigating from one end of screens to another when it would be just a flick of a thumb. Another problem it would resolve - lost cursor, to find it I move the ball, but the cursor is just hid behind the screen edge., I could just keep spinning the ball and look for the cursor move across the screens.
